piet-coregraphics/src/text.rs

```diff
@@ -35,11 +43,228 @@ pub struct CoreGraphicsTextLayout {
+    /// for the range that begins at `self.last_resolved_pos`.
+    fn current_font(&self) -> CTFont {
+        //TODO: this is where caching would happen, if we were implementing caching;
+        //store a tuple of attributes resolves to a generated CTFont.
+        unsafe {
+            let weight_key = CFString::wrap_under_create_rule(font_descriptor::kCTFontWeightTrait);
+            let weight = convert_to_coretext(self.weight.payload);
+            let family_key =
+                CFString::wrap_under_create_rule(font_descriptor::kCTFontFamilyNameAttribute);
+            let family = self.font.payload.0.family_name();
+
+            let traits_key =
+                CFString::wrap_under_create_rule(font_descriptor::kCTFontTraitsAttribute);
+            let mut traits = CFMutableDictionary::new();
+            traits.set(weight_key, weight.as_CFType());
+            if self.italic.payload {
+                let symbolic_traits_key =
+                    CFString::wrap_under_create_rule(font_descriptor::kCTFontSymbolicTrait);
+                let symbolic_traits = CFNumber::from(font_descriptor::kCTFontItalicTrait as i32);
+                traits.set(symbolic_traits_key, symbolic_traits.as_CFType());
+            }
+
+            let mut attributes = CFMutableDictionary::new();
+            attributes.set(traits_key, traits.as_CFType());
+            attributes.set(family_key, CFString::new(&family).as_CFType());
+
+            let descriptor = font_descriptor::new_from_attributes(&attributes.to_immutable());
+            font::new_from_descriptor(&descriptor, self.size.payload)
+        }
+    }
+
+    /// After we have added a span, check to see if any of our attributes are no
+    /// longer active.
+    ///
+    /// This is stateful; it requires that `self.last_resolved_pos` has been just updated
+    /// to reflect the end of the span just added.
+    fn update_after_adding_span(&mut self) {
+        let remaining_range = self.last_resolved_pos..self.text.len();
+        if self.font.range.end == self.last_resolved_pos {
+            self.font = Span::new(self.default_font.clone(), remaining_range.clone());
+        }
+
+        if self.weight.range.end == self.last_resolved_pos {
+            self.weight = Span::new(FontWeight::REGULAR, remaining_range.clone());
+        }
+
+        if self.size.range.end == self.last_resolved_pos {
+            self.size = Span::new(self.default_font.0.pt_size(), remaining_range.clone());
+        }
+
+        if self.italic.range.end == self.last_resolved_pos {
+            self.italic = Span::new(false, remaining_range);
+        }
+    }
+}
+
+/// coretext uses a float in the range -1.0..=1.0, which has a non-linear mapping
+/// to css-style weights. This is a fudge, adapted from QT:
+///
+/// https://git.sailfishos.org/mer-core/qtbase/commit/9ba296cc4cefaeb9d6c5abc2e0c0b272f2288733#1b84d1913347bd20dd0a134247f8cd012a646261_44_55
+fn convert_to_coretext(weight: FontWeight) -> CFNumber {
```

Comment:
I would have a TODO to implement piecewise linear interpolation, which I think is the strategy to minimize the error. But this is Good Enough.